tests/testsuite/update.rs

```diff
@@ -2096,3 +2095,466 @@ fn update_breaking_specific_packages() {
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [package]
+                name  =  "bar"
+                version  =  "0.0.1"
+                edition  =  "2015"
+                authors  =  []
+
+                [dependencies]
+                dep  =  "1.1"
+            "#,
+        )
+        .file("bar/src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile").run();
+
+    Package::new("dep", "1.1.1").publish();
+    Package::new("dep", "2.0.0").publish();
+
+    // Will upgrade the direct dependency
+    p.cargo("update -Zunstable-options --breaking dep@1.0")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        // FIXME: Should upgrade a dependency here.
+        .with_stderr("")
+        .run();
+
+    // But not the transitive one, because bar is not a workspace member
+    p.cargo("update -Zunstable-options --breaking dep@1.1")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr("")
+        .run();
+
+    // A non-breaking update is different, as it will update transitive dependencies
+    p.cargo("update dep@1.1")
+        .with_stderr(
+            "\
+[UPDATING] `[..]` index
+[LOCKING] 1 package to latest compatible version
+[UPDATING] dep v1.1.0 -> v1.1.1 (latest: v2.0.0)
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn update_breaking_mixed_compatibility() {
+    Package::new("mixed-compatibility", "1.0.0").publish();
+    Package::new("mixed-compatibility", "2.0.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [workspace]
+                members = ["foo", "bar"]
+            "#,
+        )
+        .file(
+            "foo/Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                mixed-compatibility = "1.0"
+            "#,
+        )
+        .file("foo/src/lib.rs", "")
+        .file(
+            "bar/Cargo.toml",
+            r#"
+                [package]
+                name = "bar"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                mixed-compatibility = "2.0"
+            "#,
+        )
+        .file("bar/src/lib.rs", "")
+        .build();
+
+    p.cargo("generate-lockfile").run();
+
+    Package::new("mixed-compatibility", "2.0.1").publish();
+
+    p.cargo("update -Zunstable-options --breaking")
+        .masquerade_as_nightly_cargo(&["update-breaking"])
+        .with_stderr(
+            "\
+[UPDATING] `[..]` index
+[UPGRADING] mixed-compatibility ^1.0 -> ^2.0
+[LOCKING] 1 package to latest compatible version
+[ADDING] mixed-compatibility v2.0.1
+",
+        )
+        .run();
+}
+
+#[cargo_test]
+/// In order to handle cases like this, it's not enough to keep an upgrade map
```

Comment:
Stylistically, I feel like documentation should go before attributes

Reply:
Moved docs elsewhere (to implementation).